staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go

```diff
@@ -96,7 +96,21 @@ func (c *DiscoveryController) sync(version schema.GroupVersion) error {
 			Version:      crd.Spec.Version,
 		})
 
-		if crd.Spec.Version != version.Version {
+		foundThisVersion := false
```

Comment:
I'm very surprised that discovery info is not updated synchronously with the handlers being added/removed. This whole thing seems like a bug, as discovery information could differ from what the server actually serves? Am I missing something?

Reply:
discovery is served from the aggregator and driven by APIService objects, which unify compiled-in local APIs, dynamic local APIs (CRDs), and remote extension APIs. Because there are two objects in play, and we don't have a multi-resource transactional API, the updates are not synchronized. 

In practice, synchronization wouldn't help clients much, since they request discovery docs, then make API requests based on them which are separated in time and outside any synchronous guarantee. 

A similar situation exists with HA API servers where you could get discovery docs from one and an API request routed to another.